crates/arroyo-connectors/src/nats/mod.rs

```diff
@@ -0,0 +1,316 @@
+            },
+            Some(other) => bail!("Unknown auth type '{}'", other),
+        };
+
+        Ok(NatsConfig {
+            authentication: nats_auth,
+            servers: nats_servers,
+        })
+    }
+
+    pub fn table_from_options(options: &mut HashMap<String, String>) -> anyhow::Result<NatsTable> {
+        let mut client_config = HashMap::new();
+        for (k, v) in options.iter() {
+            client_config.insert(k.clone(), v.clone());
+        }
+        let conn_type = pull_opt("type", options)?;
+        let nats_table_type = match conn_type.as_str() {
+            "source" => ConnectorType::Source {
+                stream: Some(pull_opt("nats.stream", options)?),
+            },
+            "sink" => ConnectorType::Sink {
+                subject: Some(pull_opt("nats.subject", options)?),
+            },
+            _ => bail!("Type must be one of 'source' or 'sink'"),
+        };
+
+        Ok(NatsTable {
+            connector_type: nats_table_type,
+            client_configs: client_config,
+        })
+    }
+}
+
+impl Connector for NatsConnector {
+    type ProfileT = NatsConfig;
+    type TableT = NatsTable;
+
+    fn name(&self) -> &'static str {
+        "nats"
+    }
+
+    fn metadata(&self) -> arroyo_rpc::api_types::connections::Connector {
+        arroyo_rpc::api_types::connections::Connector {
+            id: "nats".to_string(),
+            name: "Nats".to_string(),
+            icon: ICON.to_string(),
+            description: "Read and write from a NATS cluster".to_string(),
+            enabled: true,
+            source: true,
+            sink: true,
+            testing: true,
+            hidden: false,
+            custom_schemas: true,
+            connection_config: Some(CONFIG_SCHEMA.to_string()),
+            table_config: TABLE_SCHEMA.to_owned(),
+        }
+    }
+
+    fn config_description(&self, config: Self::ProfileT) -> String {
+        (*config.servers).to_string()
+    }
+
+    fn table_type(&self, _: Self::ProfileT, table: Self::TableT) -> ConnectionType {
+        match &table.connector_type {
+            ConnectorType::Source { .. } => ConnectionType::Source,
+            ConnectorType::Sink { .. } => ConnectionType::Sink,
+        }
+    }
+
+    fn get_schema(
+        &self,
+        _: Self::ProfileT,
+        _: Self::TableT,
+        s: Option<&ConnectionSchema>,
+    ) -> Option<ConnectionSchema> {
+        s.cloned()
+    }
+
+    fn test(
+        &self,
+        _: &str,
+        _: Self::ProfileT,
+        _: Self::TableT,
+        _: Option<&ConnectionSchema>,
+        _tx: Sender<TestSourceMessage>,
+    ) {
+        // TODO: Implement a full-fledge `NatsTester` struct for testing the client connection,
+        // the stream or subject existence and access permissions, the deserialization of messages
+        // for the specified format, the authentication to the schema registry (if any), etc.
+        let (tx, _rx) = tokio::sync::oneshot::channel();
```

Comment:
For testing, you should use the tx that's passed in to the test method rather than creating a new one (see the redis connector for a simple example of this). So basically in line 141 change that argument to `tx`, remove this line, and 148-151, and it shuold work.

Reply:
Thanks for pointing this out, looks like I introduced this when refactoring from `0.9` to `0.10`. This is fixed although a real tester still should be implemented